### ISSUE TITLE
Support for sequences of length 0 in LSTM/GRU

### ISSUE THREAD

**Issue by chrisieh** [lwtnn/lwtnn#80]
The implementation of LSTM in keras also handles input sequences that are completely masked by returning a vector of zeros. 
```
import numpy as np
from keras.layers import Input, Masking, LSTM
from keras.models import Model

masked_input = np.ones((1, 5, 4))

x = Input(shape=(5, 4))
mask = Masking(mask_value=1.0)(x)
lstm = LSTM(3, return_sequences=False)(mask)

model = Model(x, lstm)

model.predict(masked_input)
# array([[0., 0., 0.]], dtype=float32)
```
At the moment lwtnn does not allow to compute networks when passing empty input vectors. For networks with a single recurrent branch and no other inputs this feature is not needed. However, if the recurrent branch is an auxiliary input to the network it makes sense to return a zero vector to allow to compute the network output.
Given the current structure of lwtnn, would this be feasible to implement without large structural changes?

**Comment by dguest** [comment]
I think the required changes would be minimal. We prevented empty inputs before but I think the main reason was that the implementation at the time returned undefined results. I'll take a deeper look.

**Comment by dguest** [comment]
Hi @chrisieh, I think I have a fix for this, and I'll merge it in if all the tests pass. I did have to make the assumption that we wanted a vector of zeros when `return_sequences=False` and an empty sequence when `return_sequences=True`. I was too lazy to check what Keras does in the later case.

**Comment by chrisieh** [comment]
Hi @dguest, thanks for the fix. I'll give it a spin and will report back.

**Comment by dguest** [comment]
Ok, let me know. 

I made a merge request to update the ATLAS version of lwtnn a while back, which is still open. If you can verify this fix relatively soon I could update that merge request to include these improvements as well. But this would only be needed if you'd like to get this into the next analysis release; otherwise there's no urgency.

**Comment by chrisieh** [comment]
Hi @dguest, I cross-checked the output of lwtnn on empty sequences is consistent with what keras predicts. It would be very convenient if this could be merged into atlas externals. But if it's too late for that it's also not a big deal to include lwtnn in our code directly.

Thanks for your work!

**Comment by dguest** [comment]
OK, I've created a new release and updated the ATLAS-side merge request to use it. See [merge request 203 in atlas externals][1].

[1]: https://gitlab.cern.ch/atlas/atlasexternals/merge_requests/203
